### DISCUSSION TITLE
inlabru objects with terra covariates in targets

### DISCUSSION THREAD

**Discussion by dmi3kno** [inlabru-org/inlabru#217] [General]
OBS! What I am writing below is not specific to inlabru, but rather to the use of disk-stored model objects dependent on `terra`.

I have been trying to make my `inlabru` model work in targets. I have spatial covariates defined as terra rasters. Terra help says (`?terra`)

> These classes hold a C++ pointer to the data "reference class" and that creates some limitations. They cannot be recovered from a saved R session either or directly passed to nodes on a computer cluster. Generally, you should use `writeRaster()` to save SpatRaster objects to disk (and pass a filename or cell values to cluster nodes). Also, see `wrap()`.

When `inlabru` model is saved into an `rds` file, the pointer to the C++ class gets serialized. Therefore, when an `inlabru` model is read from disk in a new target, the original pointer is lost. 

I see two solutions to this:
- Always **train and predict in the same target**. Since the model object can not be saved to disk before prediction, both training and prediction should happen in the same environment, and only the predictions saved to disk. This limits the ability to predict multiple times from the same model and limits the effectiveness of task scheduling (the expensive training needs to be rerun every time the prediction is saved).
- Inlabru can be adapted to **work with compacted (wrapped) rasters**. When the raster is passed to `wrap` it becomes
```r
system.file("extdata/gorillas_cov.tif", package="inlabru") |> rast() |> wrap(proxy=FALSE)
#> [1] "This is a PackedSpatRaster object. Use 'terra::unwrap()' to unpack it"
```
This the format that can be written on disk and whatever necessary. Even if the `proxy=TRUE` is used (part of the file is cached on disk), the path to the cache will be serialized in PackedSpatRaster and, when unwrapped, restored correctly. Therefore the solution could be that `inlabru` could be taught how to handle PackedSpatRaster (and vectors). The simple logic could be twofold:
  -  When encountering a packed raster (e.g. in components or in the model object itself) it should unwrap it before proceeding
  - When saving the reference to a raster into a model object, it should wrap it (possibly with proxy caching).
 
A more complex approach could be to handle raster paths, but that seems like a stretch. Otherwise, we just let users know that the model object with `terra` references can not be written and recovered from disk.

**Comment by finnlindgren** [comment]
Are the terra objects from the global workspace actually stored in the inlabru output objects at all? The only way that could happen would be via the component and formula environments. I'd need to write some test cases to check that.
In any case it would be good to get clarity in what objects actually need to be present in which environments; currently, I believe inlabru stores multiple copies of some data, and that should be reduced.

What I would expect to be the correct approach is to load/create the terra objects before bru(), as well as before calling predict(), i.e. to call `terra::rast()` or similar in each target. In pseudo-code:
```
# Target 1, fitting:
covar <- terra::rast(...)
fit <- bru(components = ~ cov(covar, ...), lik)

# Target 2, prediction:
covar <- terra::rast(...)
pred <- predict(components, data = spatial_location_sf_object, formula = ~ ...)
```

**Reply by finnlindgren** [reply]
Where I can see a potential problem is if the `covar` object in target 1 gets associated with an environment in `fit` and the contents of that environment get stored explicitly in `fit`. Then after save/load that object would be "broken", but also risk blocking access to the fresh version created in target 2, causing the `predict()` call to fail evaluating the `cov` model component correctly.
If this is what does indeed happen, a `reprex` would be highly appreciated.

**Reply by dmi3kno** [reply]
This is my understanding. I will cook up a reprex

**Reply by finnlindgren** [reply]
Thanks. The difficulty with R variable scoping for this is when one constructs parts of the model in different environments, such as inside function calls, but need to use them in other environments. But perhaps there's a way to reduce the number of environments stored in the result object, so that it's easier to keep track of what needs to be done.  Also, this is a problem that might be an issue more generally in R, e.g. for `lm()` results objects, and I'm not sure if they have a solution either.

**Reply by dmi3kno** [reply]
The difference is that terra objects are not serializeable. This object format is relatively new in R, so it is not yet recognized as a problem.

**Reply by finnlindgren** [reply]
Yes, I meant that these types of objects would be a potential problem for all model fit objects that rely on environments. But I get your point that it's not a widely known issue yet.

**Comment by finnlindgren** [comment]
Adding support for `PackedSpatRaster` looks quite doable since the main place in the code would be to add an `eval_spatial.PackedSpatRaster()` method.
